php/libraries/TimePointData.php

```diff
@@ -0,0 +1,106 @@
+<?php declare(strict_types=1);
+/**
+ * A TimePointData object is a DTO (data transfer object) to
+ * store a representation of data associated with a TimePoint
+ * object in LORIS.
+ *
+ * The TimePointData may be partially loaded, in which case any
+ * getter trying to access an unknown will throw an exception.
+ * This allows modules to only load relevant data in a query,
+ * and avoid the overhead of the TimePoint::singleton() constructor.
+ *
+ * For instance, a query that selected CenterID and ProjectID on
+ * the database could instantiate a TimePoint as:
+ *    $visit = new TimePoint(new TimePointData(null, $projectid, $centerid));
+ *
+ * with data loaded from the query, and then call
+ * `$visit->isAccessibleBy($user)` without having to load all the
+ * over data that TimePoint::singleton() does. If an attempt is
+ * made to access a property not in the DTO (session in this example),
+ * an exception is thrown, rather than silently returning the incorrect
+ * value.
+ *
+ * @license http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ */
+class TimePointData
+{
+    /**
+     * The TimePoint's ProjectID
+     *
+     * @var ?ProjectID
+     */
+    protected $projectID;
+
+    /**
+     * The TimePoint's CenterID
+     *
+     * @var ?int
+     */
+    protected $centerID;
+
+    /**
+     * The TimePoint's SessionID
+     *
+     * @var ?SessionID
+     */
+    protected $sessionID;
+
+    /**
+     * Construct a TimePointData DTO for a TimePoint object
+     *
+     * @param ?SessionID $SessionID The SessionID for the TimePoint
+     * @param ?ProjectID $ProjectID The ProjectID for the TimePoint
+     * @param ?int       $CenterID  The CenterID for the TimePoint
+     */
+    public function __construct(
+        ?SessionID $SessionID,
+        ?ProjectID $ProjectID,
+        ?int $CenterID
```

Comment:
If these values can't be "got" when they are null, should they be able to be set as null in the first place?

Reply:
They might not be loaded from the database if your code doesn't care about them, even though in the database there's data. How would you handle that situation if they were non-nullable in the code?

Reply:
If there's situation where they can acceptably be `null` in the object, I think instead of making them non-nullable, I would just return `null` in the "getter" function rather than throw an exception. To me at least, if they can be set as `null`, then retrieving them when they are `null` should be allowed.

Reply:
Perhaps if there's a validator function that you can run the object through to make sure it is in a `valid` state, then that would throw an exception for null values that shouldn't be null after having fully processed the timepoint.

Reply:
Where would that be used?

Reply:
and I think it would need to be in the TimePointData class because the properties are protected.